scripts/upgrade.sh

```diff
@@ -0,0 +1,203 @@
+    -v | --verbose           Run in verbose mode
+EOF
+}
+
+
+TMPFILES=$(mktemp -d)
+
+cat << EOF >> "${LOGFILE}"
+--- MetalK8s Upgrade started on $(date -u -R) ---
+EOF
+
+exec > >(tee -ia "${LOGFILE}") 2>&1
+
+cleanup() {
+    rm -rf "${TMPFILES}" || true
+}
+
+trap cleanup EXIT
+
+
+run_quiet() {
+    local name=$1
+    shift 1
+
+    echo -n "> ${name}..."
+    local start
+    start=$(date +%s)
+    set +e
+    "$@" 2>&1 | tee -ia "${LOGFILE}" > "${TMPFILES}/out"
+    local RC=$?
+    set -e
+    local end
+    end=$(date +%s)
+
+    local duration=$(( end - start ))
+
+    if [ $RC -eq 0 ]; then
+        echo " done [${duration}s]"
+    else
+        echo " fail [${duration}s]"
+        cat >/dev/stderr << EOM
+
+Failure while running step '${name}'
+
+Command: $@
+
+Output:
+
+<< BEGIN >>
+EOM
+        cat "${TMPFILES}/out" > /dev/stderr
+
+        cat >/dev/stderr << EOM
+<< END >>
+
+This script will now exit
+
+EOM
+
+        exit 1
+    fi
+}
+
+run_verbose() {
+    local name=$1
+    shift 1
+
+    echo "> ${name}..."
+    "$@"
+}
+
+run() {
+    if [ "$VERBOSE" -eq 1 ]; then
+        run_verbose "${@}"
+    else
+        run_quiet "${@}"
+    fi
+}
+
+die() {
+    echo 1>&2 "$@"
+    return 1
+}
+
+__read_version() {
+  "${SALT_CALL}" slsutil.renderer string="{{ pillar.metalk8s.nodes[grains.id].version }}"
+}
+
+get_salt_container() {
+    local -r max_retries=10
+    local salt_container='' attempts=0
+
+    while [ -z "$salt_container" ] && [ $attempts -lt $max_retries ]; do
+        salt_container="$(crictl ps -q \
+            --label io.kubernetes.pod.namespace=kube-system \
+            --label io.kubernetes.container.name=salt-master \
+            --state Running)"
+        (( attempts++ ))
+    done
+
+    if [ -z "$salt_container" ]; then
+        echo "Failed to find a running 'salt-master' container" >&2
+        exit 1
+    fi
+
+    echo "$salt_container"
+}
+
+orchestrate_upgrade() {
+
+    run "Performing preliminary bootstrap node upgrade in local mode(this may take a while)" \
+    "${SALT_CALL}" --local state.sls metalk8s.roles.bootstrap saltenv=metalk8s-"$TO_VERSION" \
+        pillar="{'metalk8s': {'endpoints': $(salt-call --out txt pillar.get metalk8s:endpoints \
+        | cut -c 8-)}}"
+
+    SALT_MASTER_CALL=(crictl exec -i "$(get_salt_container)")
+
+    run "Performing Pre-Upgrade prechecks" \
+    "${SALT_MASTER_CALL[@]}" salt-run state.orchestrate metalk8s.orchestrate.upgrade.precheck saltenv=metalk8s-"$TO_VERSION" \
+        pillar="{'orchestrate': {'dest_version': '$TO_VERSION'}}"
+
+    run "Launching rolling upgrade (this may take a while)" \
+    "${SALT_MASTER_CALL[@]}" salt-run state.orchestrate metalk8s.orchestrate.upgrade saltenv=metalk8s-"$TO_VERSION" \
+        pillar="{'orchestrate': {'dest_version': '$TO_VERSION'}}"
+
+    run "Verifying that upgrade was successful(New version installed)" \
+    "${SALT_CALL}" slsutil.renderer string="{{ pillar.metalk8s.nodes[grains.id].version }}"
```

Comment:
Btw this is not "Verifying" at this will only print the version of the bootstrap node and not compare it with the destination one.

And btw you can't compare it with the destination one as during upgrade (and downgrade) some nodes may not change version.

For example if you do an upgrade to 2.2.0 and you have one node in 2.2.1 this node will stay in 2.2.1 even if destination is 2.2.0 because it's upgrade